src/crimson/osd/osd_operation.h

```diff
@@ -0,0 +1,218 @@
+
+#pragma once
+
+#include <seastar/core/shared_mutex.hh>
+#include <seastar/core/future.hh>
+
+#include <vector>
+#include <set>
+#include <boost/intrusive/list.hpp>
+#include <boost/intrusive_ptr.hpp>
+#include <boost/smart_ptr/intrusive_ref_counter.hpp>
+
+#include "common/Formatter.h"
+
+namespace ceph::osd {
+
+enum class OperationTypeCode {
+  client_request = 0,
+  peering_event,
+  compound_peering_request,
+  last_op
+};
+
+static constexpr const char* const OP_NAMES[] = {
+  "client_write",
+  "peering_event",
+  "compound_peering_request"
+};
+
+class OperationRegistry;
+
+using registry_hook_t = boost::intrusive::list_member_hook<
+  boost::intrusive::link_mode<boost::intrusive::auto_unlink>>;
+
+class Operation;
+class Blocker;
+
+template <typename... T>
+class blocking_future {
+  friend class Operation;
+  friend class Blocker;
+  Blocker *blocker = nullptr;
+  seastar::future<T...> fut;
+  blocking_future(Blocker *b, seastar::future<T...> &&f)
+    : blocker(b), fut(std::move(f)) {}
+
+  template <typename... V, typename... U>
+  friend blocking_future<V...> make_ready_blocking_future(U&&... args);
+};
+
+template <typename... V, typename... U>
+blocking_future<V...> make_ready_blocking_future(U&&... args) {
+  return blocking_future<V...>(
+    nullptr,
+    seastar::make_ready_future<V...>(std::forward<U>(args)...));
+}
+
+class Blocker {
+protected:
+  virtual void dump_detail(Formatter *f) const = 0;
+
+public:
+  template <typename... T>
+  blocking_future<T...> get_blocking_future(seastar::future<T...> &&f) {
+    return blocking_future(this, std::move(f));
+  }
+
+  void dump(Formatter *f) const;
+
+  virtual const char *get_type_name() const = 0;
+
+  virtual ~Blocker() = default;
+};
+
+template <typename T>
+class BlockerT : public Blocker {
+public:
+  const char *get_type_name() const final {
+    return T::type_name;
+  }
+
+  virtual ~BlockerT() = default;
+};
+
+class Operation : public boost::intrusive_ref_counter<
+  Operation, boost::thread_unsafe_counter> {
+  friend class OperationRegistry;
+  registry_hook_t registry_hook;
+
+  std::set<Blocker*> blockers;
+  uint64_t id = 0;
+  void set_id(uint64_t in_id) {
+    id = in_id;
+  }
+protected:
+  virtual void dump_detail(Formatter *f) const = 0;
+
+public:
+  uint64_t get_id() const {
+    return id;
+  }
+
+  virtual OperationTypeCode get_type() const = 0;
+  virtual const char *get_type_name() const = 0;
+  virtual void print(std::ostream &) const = 0;
+
+  void add_blocker(Blocker *b) {
+    blockers.insert(b);
+  }
+
+  void clear_blocker(Blocker *b) {
+    blockers.erase(b);
+  }
+
+  template <typename... T>
+  seastar::future<T...> with_blocking_future(blocking_future<T...> &&f) {
+    if (f.fut.available() || f.fut.failed()) {
+      return std::move(f.fut);
+    }
+    ceph_assert(f.blocker);
+    add_blocker(f.blocker);
+    return std::move(f.fut).then_wrapped([this, blocker=f.blocker](auto &&arg) {
+      clear_blocker(blocker);
```

Comment:
yeah, a given operation can be blocked by multiple blockers. but, i think, the previously cleared blocker could block again after the operation is blocked and unblocked by its successor blockers in the chain. put in other way, we might want to rewind and walk through the blockers, if any of the previous blockers has such a property -- its blocking-ness is subject to change over time.

Reply:
I don't think that's possible (at least not in usages so far).  I think the assumption is that a single operation will block on a particular blocker at most once at a time.  There may be more than one such blocker at any one time, but the code treats them as independent, there isn't any lifo behavior.